lawruixi iP

lawruixi opened the pull request:
# Duchess

> "If you believe it, it will happen."  – Law's Sixth Law

Duchess is defined as the wife or widow of a Duke. And [Duchess](https://github.com/lawruixi/ip/tree/master) is no exception - it is happily married to Duke. As we all know, happy individuals attract other happy individuals. This means that misery likes to collude. This also means that Duchess is basically Duke, but better. Did any of what I say make any sense? Of course not, that's not what you're here for.

You're here to see the awesome, cool, fantabulous features I have added to Duke, to turn him (her?) into Duchess.

Here's what you need to know!

- This sentence has exactly _seven_ words.
- Venus is the only planet that rotates clockwise in the the Solar System.
- **More than 84.6%** of all statistics were made up on the spot.
- The rarest blood type is ***AB negative***.
- If you multiply 21 978 by 4, it turns backwards.
- You did not notice I wrote "the the" on the second bullet point of this list.
- You just read the second bullet point again, before looking at this one.

Don't believe me? Try out Duchess yourself! Download it from [here](https://www.youtube.com/watch?v=dQw4w9WgXcQ), ~which is definitely not a YouTube link to a famous 80's song~. All you need are three simple steps:

1. Find a lost puppy.
1. Slay some world-ending deity to prevent a catastrophic calamity from descending onto our world.
1. There is no third step.

Once you have it installed, make sure you satisfy the following conditions:
- [ ] You are installing the cool Linux OS :sunglasses:.
- [ ] You are not a dog.
- [x] You think the author of this comment is intelligent, beautiful, insightful, alluring, fabulous, pretty, mystical and inspiring.

Lastly, run it by adding the following actions in your `~/.config/duchess/run.sh`:
```py
print("".join([chr(ord(c) + (1 if i % 2 == 0 else -1)) for i, c in enumerate("gfkmn!vpqmc\"")]));
```

And with that, you should be done!

<sup>Thanks for reading! Check out my other stuff [here](http://52.76.47.19/) and [here](https://lawruixi.itch.io/).</sup>

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

lawruixi committed:
renamed Duke to Duchess :D

lawruixi committed:
Fixed bug relating to Java unfamiliarity and also added the greetings functions

lawruixi committed:
added echoing functionality

lawruixi committed:
added storing text, listing stored text

lawruixi committed:
added basic Task class

lawruixi committed:
renamed Text to Tasks

lawruixi committed:
[WIP]: added regex matching for the commands and will soon add parsing for mark and unmark commands

lawruixi committed:
added parsing of mark and unmark commands, works fine without error handling as for now

lawruixi committed:
added message to confirm the mark/unmark action after doing so.

lawruixi committed:
added the three new Task subclasses

lawruixi committed:
added todo command related methods, and fixed some old documentation.

lawruixi committed:
[WIP]: added the addToDo method

lawruixi committed:
finished processing for ToDos, Deadlines and Events

also changed 1-indexing for user-inputted indexes, and fixed regex
patterns not maching numbers.

lawruixi committed:
removed three duplicate addTask methods in favour of one unified method

also, resolved a %d bug that was not string-formatted

lawruixi committed:
changed the command prompt and also the Duchess print formatting

lawruixi committed:
updated unit test case

lawruixi committed:
added "error" handling for invalid command.

lawruixi committed:
added error handling for todos and deadlines, and also aliased "exit" to
"bye"

lawruixi committed:
added events error handling, and fixed string trim issue

lawruixi committed:
updated unit test cases

lawruixi committed:
added error handling for mark and unmark

lawruixi committed:
updated test cases

lawruixi committed:
added delete command

lawruixi committed:
updated unit test cases

lawruixi committed:
now allows for saving of data between sessions

lawruixi committed:
Merge branch 'branch-Level-7'

lawruixi committed:
update deadline to use localdate

earlier, deadline class stored localdate as a string.

of course, this causes some problems with keeping track of valid dates
etc internally.

hence, this change will allow it to verify actual dates by prompting
user to enter them in yyyy-MM-dd format.

lawruixi committed:
Merge branch 'branch-Level-8'

lawruixi committed:
move parser methods into new Parser class

previously, parser methods are in Duchess class itself.

for the sake of clarity, they are now moved into Parser class.

lawruixi committed:
rename Data to Storage

renamed Data.java to Storage.java, as per requirements in the task.

lawruixi committed:
add aliases to "list" and "delete" commands

previously, using bash-like syntax "ls" or "rm" did not do the
respective list or delete commands, despite that being very natural for
Linux users to navigate its interface.

this change adds aliases to the list and delete commands such that "ls"
will call "list" and "rm" will call "delete".

lawruixi committed:
move methods to Ui.java

previously, UI methods were in Duchess.java.

moved methods to Ui.java as per requirements. Ui is still a static class
right now as I don't really see the need for a spearate UI object.

lawruixi committed:
create TaskList class

changed ArrayList<Task> to TaskList class as per specifications.

lawruixi committed:
Merge branch 'more-oop'

lawruixi committed:
put the files in one package

previously, the files were not in any package.

now, the files are in a package as per specifications.

lawruixi committed:
Merge branch 'A-Packages'

lawruixi committed:
Merge remote-tracking branch 'origin/add-gradle-support' into A-Gradle

lawruixi committed:
update gradle for Duchess

gradle is now updated and renamed to fit this project's package structure and new name.

lawruixi committed:
Merge branch 'A-Gradle'

lawruixi committed:
add unit tests for TaskList and Utility

added unit tests for the classes TaskList and Utility.

lawruixi committed:
update to fit coding standard

lawruixi committed:
added javadocs for thrown exceptions

lawruixi committed:
added search function and a bunch of utility functions to tasklist

lawruixi committed:
Merge branch 'A-CodingStandard'

lawruixi committed:
Merge branch 'Level-9'

lawruixi committed:
add checkstyle files and checkstyle plugins

lawruixi committed:
Merge branch 'branch-A-Checkstyle'

lawruixi committed:
updated coding style conventions to the test files

lawruixi committed:
added basic code for duchess gui

lawruixi committed:
added GUI for Duchess

lawruixi committed:
Merge branch 'branch-Level-10'

wjayee commented on src/main/java/duchess/TaskList.java:
```diff
@@ -0,0 +1,164 @@
+ * Represents a list of tasks. This is a wrapper around a ArrayList&lt;Task&gt;.
+ */
+public class TaskList {
+    private ArrayList<Task> tasks;
```
Nice naming of the arraylist !

wjayee commented on src/main/java/duchess/TaskList.java:
```diff
@@ -0,0 +1,164 @@
+     *
+     * @param tasksArrayList - an ArrayList of Tasks.
+     */
+    TaskList(ArrayList<Task> tasksArrayList) {
```
Perhaps you could add access modifiers for your constructors ?

wjayee commented on src/main/java/duchess/Duchess.java:
```diff
@@ -0,0 +1,254 @@
+        Scanner sc = new Scanner(System.in);
+        String userInput = "";
+
+        while (!Parser.isExitCommand(userInput)) {
```
I like your indentation and spacing here !

wjayee commented on src/main/java/duchess/Storage.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    /**
+     * Saves the specified ArrayList of tasks to the data file specified by DATA_DIRECTORY and DATA_FILE.
```
I like how your comment clearly explains what this method does !

lynnlow175 commented on src/main/java/duchess/Deadline.java:
```diff
@@ -0,0 +1,88 @@
+     */
+    public static Deadline fromSaveString(String s) {
+        String[] splitString = s.split(Task.SAVE_STRING_DELIMITER);
+        // Not enough arguments; minmally, it needs the Type, the Marked status, the Name and the Deadline.
```
Good that you explained your code that might be otherwise hard to understand! Good job:)

lynnlow175 commented on src/main/java/duchess/Duchess.java:
```diff
@@ -0,0 +1,254 @@
+     * @param s - the string to be stored.
+     */
+    private static void listTasks() {
+        Ui.duchessPrint("Here are the things you said!! ヽ(^o^)丿");
```
Perhaps, you could put this print statement under the Ui class? It kind of defeats the point when you still have to type out the string here :(

lynnlow175 commented on src/main/java/duchess/Duchess.java:
```diff
@@ -0,0 +1,254 @@
+                continue;
+            }
+
+            Ui.duchessPrint("(´；ω；`) Oopsies... I don't know what that means ;-;");
```
I love these cute messages with emoticons!

lynnlow175 commented on src/main/java/duchess/Duchess.java:
```diff
@@ -0,0 +1,254 @@
+    }
+
+    /**
+     * Unmarks a task.
```
Wrong description for this method. :((

lynnlow175 reviewed (commented):
Overall, code is tidy and clear with some minor mistakes to fix. Really love the customised messages with emoticons:)

lawruixi committed:
Add assertions in code

Previously, the code used no assertions. If a method call to some
methods had invalid parameters (like null parameters), it may cause
unknown catastrophic consequences.

Now, assertions have been added to some methods. If the method call
passes invalid parameters, the method will throw an AssertionError
instead.

lawruixi committed:
removed redundant echo method

lawruixi committed:
Merge pull request #1 from lawruixi/branch-A-Assertions

Add assertions in code

lawruixi committed:
Merge branch 'master' into branch-A-CodeQuality

lawruixi committed:
Merge pull request #2 from lawruixi/branch-A-CodeQuality

removed redundant echo method

lawruixi committed:
[WIP] added tag command, tag parsing, adding of tags; currently working on formatting the tags when listing, or deleting tags

lawruixi committed:
[WIP] added tag command, tag parsing, adding of tags; currently working on formatting the tags when listing, or deleting tags

lawruixi committed:
added tags formatting

lawruixi committed:
added tags formatting

lawruixi committed:
overhauled the entire Duchess system to focus on using callback handlers
instead.

Previously, Duchess and DuchessGUI were both separate classes each with
their own versions of the same method to handle the different cases of
CLI Duchess and GUI Duchess. However, with this change, they now both
rely on callback handlers to differentiate the two cases.

This means that Duchess and DuchessGUI no longer need to be separate
classes, but can each both have a Duchess instance running the same
commands, the only difference being the callback handler.

lawruixi committed:
updated DuchessGUI too, now works

lawruixi committed:
fixed bug with 0 and 1 indexing for tagging

lawruixi committed:
fixed bug where DuchessGUI would not exit upon typing an exit command

lawruixi committed:
Merge branch 'branch-C-Tagging'

lawruixi committed:
added tag saving for deadline only

lawruixi committed:
add saving for all the task subclasses

lawruixi committed:
implement removing of tasks

lawruixi committed:
Merge branch 'branch-C-Tagging'

lawruixi committed:
made UI slightly better

lawruixi committed:
removed white border around dialog box because it doesnt play nice with rounded corners

lawruixi committed:
fixed some tag-related bugs

lawruixi committed:
fix bug involving loading Event data

lawruixi committed:
fixed bug where rm would crash upon giving too many arguments

lawruixi committed:
add docs and update README for this project.

lawruixi committed:
Update README.md

fixed a bug with the detailed acknowledgements section

lawruixi committed:
[wip] updating the commands to be their own classes

lawruixi committed:
mark task has been converted

lawruixi committed:
updated unmark task command

lawruixi committed:
updated delete command

lawruixi committed:
updated search command

lawruixi committed:
updated todo, deadline, event commands

lawruixi committed:
updated add and remove tag commands

lawruixi committed:
added the unrecognized command

lawruixi committed:
updated run.sh

lawruixi committed:
updated README to include instructions on exiting

lawruixi committed:
fixed a bug with the duchess formatting in the unrecognized command.

lawruixi committed:
update files to comply with checkstyle
